### PR TITLE
Feat/14 auth route

### LINKED ISSUE #14: Implement the auth route
**Is your feature request related to a problem? Please describe.**
We need a route to authenticate the user and retrieve a session token.

**Describe the solution you'd like**
A route that I can perform a GET *serverHost*/login and receive a token at *clientHost*/auth/**:token**

**Describe alternatives you've considered**
I implemented an approach based on the one in 'Additional context' section. My approach uses GitHub but it needs more enhancement. The approach: [social-login-github](https://github.com/jadsonluan/social-login-github)

**Additional context**
[Similiar route](https://github.com/calluswhatyouwant/musicritic/blob/master/server/spotify/spotifyAuthController.js)

**Related**
#12

### PR DESCRIPTION
**Pull request by jadsonluan**
# Description
Implement the authentication.

Fixes #14 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Clone the [client](https://github.com/jadsonluan/eu-aqui-github/) and run it (follow the [read me](https://github.com/jadsonluan/eu-aqui-github/blob/master/README.md)). 
Create a .env file (follow the .env.example that is inside of the branch).
Run the server (follow this repo readme)

- [ ] Open http://localhost:8080 and press the 'Authorize' button, then authorize the permissions in the window opened when you clicked the button. Then you'll be redirected to the profile page.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/githubAuth/route.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+router.get('/callback', (req, res) => {
+  console.log('/callback');
+  const storedState = req.cookies ? req.cookies[stateKey] : null;
+  const { state } = req.query;
+  
+  if (state && storedState === state) {
+    requestAccessToken(req, res);
+  } else {
+    const queryString = querystring.stringify({ error: 'state_mismatch' });
+    const url = getClientURL(config.client.errorPath + '?' + queryString);
```
Comment (fanny) [inline src/githubAuth/route.js]: Hummm, you could use template strings of es6
```suggestion
    const url = getClientURL(`${config.client.errorPath}?${queryString}`);
```

File: src/githubAuth/route.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      client_secret: config.github.clientSecret
+    }
+  }
+
+  request.post(requestOptions, (error, response, body) => {
+    let url;
+
+    if (!error & response.statusCode === 200) {
+      const { access_token } = body;
+      url = getClientURL(config.client.successPath + '/' + access_token)
```
Comment (fanny) [inline src/githubAuth/route.js]:
```suggestion
      url = getClientURL(`${config.client.successPath}/${access_token}`)
```

File: src/githubAuth/route.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    } else {
+      const queryString = querystring.stringify({ error: 'invalid_token' });
+      url = getClientURL(config.client.errorPath + '?' + queryString);
```
Comment (fanny) [inline src/githubAuth/route.js]:
```suggestion
      url = getClientURL(`${config.client.errorPath}?${queryString}`);
```

**Comment by jadsonluan** [comment]
I renamed the directory but i will open another PR for the config.js because we have an the #12  issue for that

File: src/auth/route.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  generateRandomState, 
+  getAuthBaseURL,
+  getClientURL
+} from './util';
+
+const router = express.Router();
+const stateKey = 'github-auth-state';
+
+router.get('/login', (req, res) => {
+  console.log('/login');
```
Comment (fanny) [inline src/auth/route.js]: Remove console.log

File: src/auth/route.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    redirect_uri: config.github.redirectUrl,
+    state
+  });
+  
+  const url = getAuthBaseURL('/authorize?' + queryString);
+  res.redirect(url);
+});
+
+router.get('/callback', (req, res) => {
+  console.log('/callback');
```
Comment (fanny) [inline src/auth/route.js]: Remove console.log

File: src/auth/route.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  const state = generateRandomState(16);
+  res.cookie(stateKey, state);
+  
+  const queryString = querystring.stringify({
+    client_id: config.github.clientId,
+    redirect_uri: config.github.redirectUrl,
+    state
+  });
+  
+  const url = getAuthBaseURL('/authorize?' + queryString);
```
Comment (fanny) [inline src/auth/route.js]: use es6 template string
```suggestion
  const url = getAuthBaseURL(`/authorize?${queryString}`);

File: src/auth/route.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+router.get('/callback', (req, res) => {
+  console.log('/callback');
+  const storedState = req.cookies ? req.cookies[stateKey] : null;
+  const { state } = req.query;
+  
+  if (state && storedState === state) {
+    requestAccessToken(req, res);
+  } else {
+    const queryString = querystring.stringify({ error: 'state_mismatch' });
+    const url = getClientURL(config.client.errorPath + '?' + queryString);
```
Comment (fanny) [inline src/auth/route.js]:
```suggestion
    const url = getClientURL(`config.client.errorPath?${queryString});
```

File: src/auth/route.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    res.redirect(url);
+  }
+});
+
+const requestAccessToken = (req, res) => {
+  console.log('requestAccessToken()');
```
Comment (fanny) [inline src/auth/route.js]: Remove console

**Pull request merged**